File: python/findleavesofbinarytree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findLeaves(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        res = []
        while root:
            self.tmp = []
            if self.isLeave(root):
                root = None
            res.append(self.tmp[:])
        return res
        
    def isLeave(self, root):
        if not root.left and not root.right:
            self.tmp.append(root.val)
            return True
        if root.left:
            if self.isLeave(root.left):
                root.left = None
        if root.right:
            if self.isLeave(root.right):
                root.right = None
        return False
            

File: java/FindLeavesOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    public List<List<Integer>> findLeaves(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        helper(root, res);
        return res;
    }
    
    private int helper(TreeNode root, List<List<Integer>> res) {
        if (root == null) {
            return 0;
        }
        
        int height = Math.max(helper(root.left, res), helper(root.right, res)) + 1;
        if (height - 1 >= res.size()) {
            res.add(new ArrayList<>());
        }
        res.get(height-1).add(root.val);
        root.left = root.right = null;
        
        return height;
    }
}
